### PR TITLE
Massive refactor

### PR DESCRIPTION
**Pull request by paddycarver**
A massive refactoring effort to:

* Change how we're serving providers (#215)
* Change how we're structuring reflection (#66, #84)
* Change where we convert from `tftypes.Value` to `attr.Value` (#172)

and clean up a few other things along the way. It turns out when you start poking at one of these, the whole ball of yarn starts unraveling. Especially since this is an offshoot of the work I was doing in #52.

### CODE REVIEW

**Comment by paddycarver** [comment]
Oh, also, this refactors Plan/Config/State to use new types, hopefully cutting down on the amount of code needed. I deleted the plan/state/config tests, but they'll be replaced with non-duplicated tests for the `ReadOnlyData` and `*Data` types. I largely lumped this in because the thought of updating those 10,000 lines of tests made me sad.

File: tfsdk/data.go [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+func setAttributeValue(ctx context.Context, path *tftypes.AttributePath, object types.Object, val attr.Value) (types.Object, diag.Diagnostics) {
+	// TODO: update object with val at path
+	return types.Object{}, nil
+}
+
+func (d *Data) Set(ctx context.Context, val interface{}, options ...SetDataOption) diag.Diagnostics {
+	return nil
+}
+
+func (d *Data) RemoveResource(ctx context.Context) {
```
Comment (bflad) [inline tfsdk/data.go:111]: Drive-by note: In principle, these new "data" handling types feel appropriate to centralize the data logic of the Config/Plan/State handling. However, I'm not sure that we should necessarily get away from having `Config`, `Plan`, and `State` types. This `RemoveResource` method is a good example of something that should only be implemented with `State` and not `Plan`.

I'm wondering if `Config`, `Plan`, and `State` should still exist with encapsulating the underlying data types, e.g. briefly

```go
type Config struct {
  ReadOnlyData
}

type Plan struct {
  *Data
}

type State struct {
  *Data
}

func (s State) RemoveResource(ctx context.Context) {/* ... */}
```

I'm also wondering if this particular change could be done outside the larger refactoring.

File: tfsdk/serve_config.go [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+import (
+	"context"
+
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+	"github.com/hashicorp/terraform-plugin-go/tfprotov6"
+)
+
+// parseConfig turns a *tfprotov6.DynamicValue and a Schema into a Config.
+func parseConfig(ctx context.Context, dv *tfprotov6.DynamicValue, schema Schema) (ReadOnlyData, diag.Diagnostics) {
```
Comment (bflad) [inline tfsdk/serve_config.go:11]: Drive-by nit: I'd be tempted to name this `newReadOnlyData` for clarity, unless its updated to return a `Config` (which I'd still opt for the new/New prefix).

File: tfsdk/serve_resource.go [hunk 3]
```diff
@@ -0,0 +1,50 @@
+package tfsdk
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+)
+
+func serveGetResourceType(ctx context.Context, provider Provider, typ string) (ResourceType, diag.Diagnostics) {
```
Comment (bflad) [inline tfsdk/serve_resource.go:10]: Drive-by question: Is there a particular reason these aren't unexported methods on `Provider`? e.g.

```go
func (p Provider) getResourceType(ctx context.Context, typ string) (ResourceType, diag.Diagnostics) {/* ... */}
```

**Review by bflad — COMMENTED** [review]
Just leaving some drive-by notes while this is in draft. 👍  It'd be super helpful to understand the desired end state (of course a little difficult while prototyping 🙁 ) and which pieces we can potentially be split out to reduce design review burden.

**Comment by bflad** [comment]
There's been _dozens_ of pull requests to accomplish even part of this. 😛 This has really served as a valuable reference for various ideas though, so thank you.
